community/primitive-collections/src/main/java/org/neo4j/concurrent/WorkSync.java

```diff
@@ -158,29 +191,127 @@ private W combine( WorkUnit<Material,W> batch )
             {
                 result = result.combine( batch.work );
             }
-            batch = batch.next;
+
+            do
+            {
+                batch = batch.next;
+            }
+            while ( batch.isCancelled() );
         }
         return result;
     }
 
     private void markAsDone( WorkUnit<Material,W> batch )
     {
+        // We will mark the first work unit as the successor - the thread in charge of unparking
+        // the other blocked threads - before we begin marking the units as done.
+        // This way, if the successor wakes up before we finish marking the units as done, it will
+        // know that it should wait and complete its task. In fact, it may trail our done-marking
+        // process and unpark threads in parallel.
```

Comment:
This whole "successor" business was not included in #7822, but strangely this comment was.
Digging through the commits, the "successor" logic was introduced in _Make WorkSync implement its own locking_, and then removed in _Optimisations and asynchronous apply method for WorkSync_.
Perhaps those same optimizations could be applied here, because the "successor" stuff is quite complex (quite possibly overly complex), and at best under documented (more comments please!), at worst unnecessary.